scalafmt-core/shared/src/main/scala/org/scalafmt/util/MarkdownFile.scala

```diff
@@ -0,0 +1,133 @@
+package org.scalafmt.util
+
+import scala.meta.inputs.Position
+import scala.meta.inputs.Input
+import scala.collection.mutable
+
+final case class MarkdownFile(input: Input, parts: List[MarkdownPart]) {
+  private val appends = mutable.ListBuffer.empty[String]
+  def appendText(text: String): Unit = {
+    appends += text
+  }
+  def renderToString: String = {
+    val out = new StringBuilder()
+    parts.foreach(_.renderToString(out))
+    appends.foreach(a => out.append(a).append("\n"))
+    out.toString()
+  }
+}
+object MarkdownFile {
+  sealed abstract class State
+  object State {
+    case class CodeFence(start: Int, backticks: String, info: String)
+        extends State
+    case object Text extends State
+  }
+  class Parser(input: Input) {
+    private val text = input.text
+    private def newPos(start: Int, end: Int): Position = {
+      Position.Range(input, start, end)
+    }
+    private def newText(start: Int, end: Int): Text = {
+      val adaptedEnd = math.max(start, end)
+      val part = Text(text.substring(start, adaptedEnd))
+      part.pos = newPos(start, adaptedEnd)
+      part
+    }
+    private def newCodeFence(
+        state: State.CodeFence,
+        backtickStart: Int,
+        backtickEnd: Int
+    ): CodeFence = {
+      val open = newText(state.start, state.start + state.backticks.length())
+      val info = newText(open.pos.end, open.pos.end + state.info.length())
+      val adaptedBacktickStart = math.max(0, backtickStart - 1)
+      val body = newText(info.pos.end, adaptedBacktickStart)
+      val close = newText(adaptedBacktickStart, backtickEnd)
+      val part = CodeFence(open, info, body, close)
+      part.pos = newPos(state.start, backtickEnd)
+      part
+    }
+    def acceptParts(): List[MarkdownPart] = {
+      var state: State = State.Text
+      val parts = mutable.ListBuffer.empty[MarkdownPart]
+      var curr = 0
+      text.linesWithSeparators.foreach { line =>
+        val end = curr + line.length()
+        state match {
+          case State.Text =>
+            if (line.startsWith("```")) {
+              val backticks = line.takeWhile(_ == '`')
+              val info = line.substring(backticks.length())
+              state = State.CodeFence(curr, backticks, info)
+            } else {
+              parts += newText(curr, end)
+            }
+          case s: State.CodeFence =>
+            if (
+              line.startsWith(s.backticks) &&
```

Comment:
this is, by the way, an example where looking for ` ```scala mdoc ` would be incorrect, since it could be nested inside some other ` ``` ` fence.

Reply:
I'm going to address the simpler points that you made and then target nested fences in a follow-up commit, since I really have no idea how hard it will be to handle those properly. 
I'll make sure to create the relevant test cases as part of that work.

Reply:
I added some tests that confirm that this is able to handle (Ignore) nested fences.